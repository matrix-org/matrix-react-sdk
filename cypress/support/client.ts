/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/// <reference types="cypress" />

import type {
    MatrixClient,
    Room,
    IContent,
    FileType,
    Upload,
    UploadOpts,
    ICreateRoomOpts,
    ISendEventResponse,
} from "matrix-js-sdk/src/matrix";
import Chainable = Cypress.Chainable;

declare global {
    // eslint-disable-next-line @typescript-eslint/no-namespace
    namespace Cypress {
        interface Chainable {
            /**
             * Returns the MatrixClient from the MatrixClientPeg
             */
            getClient(): Chainable<MatrixClient | undefined>;
            /**
             * Create a room with given options.
             * @param options the options to apply when creating the room
             * @return the ID of the newly created room
             */
            createRoom(options: ICreateRoomOpts): Chainable<string>;
            /**
             * Create a space with given options.
             * @param options the options to apply when creating the space
             * @return the ID of the newly created space (room)
             */
            createSpace(options: ICreateRoomOpts): Chainable<string>;
            /**
             * Invites the given user to the given room.
             * @param roomId the id of the room to invite to
             * @param userId the id of the user to invite
             */
            inviteUser(roomId: string, userId: string): Chainable<{}>;
            /**
             * Sets account data for the user.
             * @param type The type of account data.
             * @param data The data to store.
             */
            setAccountData(type: string, data: object): Chainable<{}>;
            /**
             * @param {string} roomId
             * @param {string} threadId
             * @param {string} eventType
             * @param {Object} content
             * @return {module:http-api.MatrixError} Rejects: with an error response.
             */
            sendEvent(
                roomId: string,
                threadId: string | null,
                eventType: string,
                content: IContent,
            ): Chainable<ISendEventResponse>;
            /**
             * @param {string} name
             * @param {module:client.callback} callback Optional.
             * @return {Promise} Resolves: {} an empty object.
             * @return {module:http-api.MatrixError} Rejects: with an error response.
             */
            setDisplayName(name: string): Chainable<{}>;
            /**
             * @param {string} url
             * @param {module:client.callback} callback Optional.
             * @return {Promise} Resolves: {} an empty object.
             * @return {module:http-api.MatrixError} Rejects: with an error response.
             */
            setAvatarUrl(url: string): Chainable<{}>;
            /**
             * Upload a file to the media repository on the homeserver.
             *
             * @param {object} file The object to upload. On a browser, something that
             *   can be sent to XMLHttpRequest.send (typically a File).  Under node.js,
             *   a a Buffer, String or ReadStream.
             */
            uploadContent(file: FileType, opts?: UploadOpts): Chainable<Awaited<Upload["promise"]>>;
            /**
             * Turn an MXC URL into an HTTP one. <strong>This method is experimental and
             * may change.</strong>
             * @param {string} mxcUrl The MXC URL
             * @param {Number} width The desired width of the thumbnail.
             * @param {Number} height The desired height of the thumbnail.
             * @param {string} resizeMethod The thumbnail resize method to use, either
             * "crop" or "scale".
             * @param {Boolean} allowDirectLinks If true, return any non-mxc URLs
             * directly. Fetching such URLs will leak information about the user to
             * anyone they share a room with. If false, will return null for such URLs.
             * @return {?string} the avatar URL or null.
             */
            mxcUrlToHttp(
                mxcUrl: string,
                width?: number,
                height?: number,
                resizeMethod?: string,
                allowDirectLinks?: boolean,
            ): string | null;
            /**
             * Gets the list of DMs with a given user
             * @param userId The ID of the user
             * @return the list of DMs with that user
             */
            getDmRooms(userId: string): Chainable<string[]>;
            /**
             * Joins the given room by alias or ID
             * @param roomIdOrAlias the id or alias of the room to join
             */
            joinRoom(roomIdOrAlias: string): Chainable<Room>;
        }
    }
}

Cypress.Commands.add("getClient", (): Chainable<MatrixClient | undefined> => {
    return cy.window({ log: false }).then((win) => win.mxMatrixClientPeg.matrixClient);
});

Cypress.Commands.add("getDmRooms", (userId: string): Chainable<string[]> => {
    return cy
        .getClient()
        .then((cli) => cli.getAccountData("m.direct")?.getContent<Record<string, string[]>>())
        .then((dmRoomMap) => dmRoomMap[userId] ?? []);
});

Cypress.Commands.add("createRoom", (options: ICreateRoomOpts): Chainable<string> => {
    return cy.window({ log: false }).then(async (win) => {
        const cli = win.mxMatrixClientPeg.matrixClient;
        const resp = await cli.createRoom(options);
        const roomId = resp.room_id;

        if (!cli.getRoom(roomId)) {
            await new Promise<void>((resolve) => {
                const onRoom = (room: Room) => {
                    if (room.roomId === roomId) {
                        cli.off(win.matrixcs.ClientEvent.Room, onRoom);
                        resolve();
                    }
                };
                cli.on(win.matrixcs.ClientEvent.Room, onRoom);
            });
        }

        return roomId;
    });
});

Cypress.Commands.add("createSpace", (options: ICreateRoomOpts): Chainable<string> => {
    return cy.createRoom({
        ...options,
        creation_content: {
            type: "m.space",
        },
    });
});

Cypress.Commands.add("inviteUser", (roomId: string, userId: string): Chainable<{}> => {
    return cy.getClient().then(async (cli: MatrixClient) => {
        const res = await cli.invite(roomId, userId);
        Cypress.log({ name: "inviteUser", message: `sent invite in ${roomId} for ${userId}` });
        return res;
    });
});

Cypress.Commands.add("setAccountData", (type: string, data: object): Chainable<{}> => {
    return cy.getClient().then(async (cli: MatrixClient) => {
        return cli.setAccountData(type, data);
    });
});

Cypress.Commands.add(
    "sendEvent",
    (roomId: string, threadId: string | null, eventType: string, content: IContent): Chainable<ISendEventResponse> => {
        return cy.getClient().then(async (cli: MatrixClient) => {
            return cli.sendEvent(roomId, threadId, eventType, content);
        });
    },
);

Cypress.Commands.add("setDisplayName", (name: string): Chainable<{}> => {
    return cy.getClient().then(async (cli: MatrixClient) => {
        return cli.setDisplayName(name);
    });
});

Cypress.Commands.add("uploadContent", (file: FileType, opts?: UploadOpts): Chainable<Awaited<Upload["promise"]>> => {
    return cy.getClient().then(async (cli: MatrixClient) => {
        return cli.uploadContent(file, opts);
    });
});

Cypress.Commands.add("setAvatarUrl", (url: string): Chainable<{}> => {
    return cy.getClient().then(async (cli: MatrixClient) => {
        return cli.setAvatarUrl(url);
    });
});

Cypress.Commands.add("joinRoom", (roomIdOrAlias: string): Chainable<Room> => {
    return cy.getClient().then((cli) => cli.joinRoom(roomIdOrAlias));
});
