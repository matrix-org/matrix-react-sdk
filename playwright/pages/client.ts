/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { JSHandle, Page } from "@playwright/test";
import { PageFunctionOn } from "playwright-core/types/structs";

import type {
    IContent,
    ICreateRoomOpts,
    ISendEventResponse,
    MatrixClient,
    Room,
    IRoomDirectoryOptions,
} from "matrix-js-sdk/src/matrix";

export class Client {
    protected client: JSHandle<MatrixClient>;

    protected getClientHandle(): Promise<JSHandle<MatrixClient>> {
        return this.page.evaluateHandle(() => window.mxMatrixClientPeg.get());
    }

    public async prepareClient(): Promise<JSHandle<MatrixClient>> {
        if (!this.client) {
            this.client = await this.getClientHandle();
        }
        return this.client;
    }

    public constructor(protected readonly page: Page) {
        page.on("framenavigated", async () => {
            this.client = null;
        });
    }

    public evaluate<R, Arg, O extends MatrixClient = MatrixClient>(
        pageFunction: PageFunctionOn<O, Arg, R>,
        arg: Arg,
    ): Promise<R>;
    public evaluate<R, O extends MatrixClient = MatrixClient>(
        pageFunction: PageFunctionOn<O, void, R>,
        arg?: any,
    ): Promise<R>;
    public async evaluate<T>(fn: (client: MatrixClient) => T, arg?: any): Promise<T> {
        await this.prepareClient();
        return this.client.evaluate(fn, arg);
    }

    /**
     * @param roomId ID of the room to send the event into
     * @param threadId ID of the thread to send into or null for main timeline
     * @param eventType type of event to send
     * @param content the event content to send
     */
    public async sendEvent(
        roomId: string,
        threadId: string | null,
        eventType: string,
        content: IContent,
    ): Promise<ISendEventResponse> {
        const client = await this.prepareClient();
        return client.evaluate(
            async (client, { roomId, threadId, eventType, content }) => {
                return client.sendEvent(roomId, threadId, eventType, content);
            },
            { roomId, threadId, eventType, content },
        );
    }

    /**
     * Send a message as a bot into a room
     * @param roomId ID of the room to send the message into
     * @param content the event content to send
     */
    public async sendMessage(roomId: string, content: IContent): Promise<ISendEventResponse> {
        const client = await this.prepareClient();
        return client.evaluate(
            (client, { roomId, content }) => {
                return client.sendMessage(roomId, content);
            },
            {
                roomId,
                content,
            },
        );
    }

    /**
     * Create a room with given options.
     * @param options the options to apply when creating the room
     * @return the ID of the newly created room
     */
    public async createRoom(options: ICreateRoomOpts): Promise<string> {
        const client = await this.prepareClient();
        return await client.evaluate(async (cli, options) => {
            const resp = await cli.createRoom(options);
            const roomId = resp.room_id;
            if (!cli.getRoom(roomId)) {
                await new Promise<void>((resolve) => {
                    const onRoom = (room: Room) => {
                        if (room.roomId === roomId) {
                            cli.off(window.matrixcs.ClientEvent.Room, onRoom);
                            resolve();
                        }
                    };
                    cli.on(window.matrixcs.ClientEvent.Room, onRoom);
                });
            }
            return roomId;
        }, options);
    }

    /**
     * Create a space with given options.
     * @param options the options to apply when creating the space
     * @return the ID of the newly created space (room)
     */
    public async createSpace(options: ICreateRoomOpts): Promise<string> {
        return this.createRoom({
            ...options,
            creation_content: {
                ...options.creation_content,
                type: "m.space",
            },
        });
    }

    /**
     * Joins the given room by alias or ID
     * @param roomIdOrAlias the id or alias of the room to join
     */
    public async joinRoom(roomIdOrAlias: string): Promise<void> {
        const client = await this.prepareClient();
        await client.evaluate(async (client, roomIdOrAlias) => {
            return await client.joinRoom(roomIdOrAlias);
        }, roomIdOrAlias);
    }

    /**
     * Make this bot join a room by name
     * @param roomName Name of the room to join
     */
    public async joinRoomByName(roomName: string): Promise<void> {
        const client = await this.prepareClient();
        await client.evaluate(
            (client, { roomName }) => {
                const room = client.getRooms().find((r) => r.getDefaultRoomName(client.getUserId()) === roomName);
                if (room) {
                    return client.joinRoom(room.roomId);
                }
                throw new Error(`Bot room join failed. Cannot find room '${roomName}'`);
            },
            {
                roomName,
            },
        );
    }

    /**
     * Invites the given user to the given room.
     * @param roomId the id of the room to invite to
     * @param userId the id of the user to invite
     */
    public async inviteUser(roomId: string, userId: string): Promise<void> {
        const client = await this.prepareClient();
        await client.evaluate((client, { roomId, userId }) => client.invite(roomId, userId), {
            roomId,
            userId,
        });
    }

    public async publicRooms(options?: IRoomDirectoryOptions): ReturnType<MatrixClient["publicRooms"]> {
        const client = await this.prepareClient();
        return await client.evaluate((client, options) => {
            return client.publicRooms(options);
        }, options);
    }
}
