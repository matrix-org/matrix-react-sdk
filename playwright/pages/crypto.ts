/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { APIRequestContext, Page, expect } from "@playwright/test";

import { HomeserverInstance } from "../plugins/homeserver";

export class Crypto {
    public constructor(
        private page: Page,
        private homeserver: HomeserverInstance,
        private request: APIRequestContext,
    ) {}

    /**
     * Check that the user has published cross-signing keys, and that the user's device has been cross-signed.
     */
    public async assertDeviceIsCrossSigned(): Promise<void> {
        const { userId, deviceId, accessToken } = await this.page.evaluate(() => ({
            userId: window.mxMatrixClientPeg.get().getUserId(),
            deviceId: window.mxMatrixClientPeg.get().getDeviceId(),
            accessToken: window.mxMatrixClientPeg.get().getAccessToken(),
        }));

        const res = await this.request.post(`${this.homeserver.config.baseUrl}/_matrix/client/v3/keys/query`, {
            headers: { Authorization: `Bearer ${accessToken}` },
            data: { device_keys: { [userId]: [] } },
        });
        const json = await res.json();

        // there should be three cross-signing keys
        expect(json.master_keys[userId]).toHaveProperty("keys");
        expect(json.self_signing_keys[userId]).toHaveProperty("keys");
        expect(json.user_signing_keys[userId]).toHaveProperty("keys");

        // and the device should be signed by the self-signing key
        const selfSigningKeyId = Object.keys(json.self_signing_keys[userId].keys)[0];

        expect(json.device_keys[userId][deviceId]).toBeDefined();

        const myDeviceSignatures = json.device_keys[userId][deviceId].signatures[userId];
        expect(myDeviceSignatures[selfSigningKeyId]).toBeDefined();
    }
}
