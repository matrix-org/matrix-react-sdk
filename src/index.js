/*
Copyright 2015, 2016 OpenMarket Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

var Skinner = require('./Skinner');

var t = require('counterpart');

t.registerTranslations('en', require('../languages/en'));
t.registerTranslations('pt-BR', require('../languages/pt-BR'));

t.setFallbackLocale('en');
t.setLocale(window.navigator.languages[0]);



module.exports.loadSkin = function(skinObject) {
    Skinner.load(skinObject);
};

module.exports.resetSkin = function() {
    Skinner.reset();
};

module.exports.getComponent = function(componentName) {
    return Skinner.getComponent(componentName);
};
