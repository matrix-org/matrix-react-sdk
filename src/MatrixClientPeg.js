/*
Copyright 2015, 2016 OpenMarket Ltd
Copyright 2017 Vector Creations Ltd.
Copyright 2017 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { MatrixClient, MemoryStore } from "matrix-js-sdk";

import utils from "matrix-js-sdk/lib/utils";
import EventTimeline from "matrix-js-sdk/lib/models/event-timeline";
import EventTimelineSet from "matrix-js-sdk/lib/models/event-timeline-set";
import sdk from "./index";
import createMatrixClient from "./utils/createMatrixClient";
import SettingsStore from "./settings/SettingsStore";
import MatrixActionCreators from "./actions/MatrixActionCreators";
import Modal from "./Modal";
import { verificationMethods } from "matrix-js-sdk/lib/crypto";
import MatrixClientBackedSettingsHandler from "./settings/handlers/MatrixClientBackedSettingsHandler";
import * as StorageManager from "./utils/StorageManager";
import IdentityAuthClient from "./IdentityAuthClient";

interface MatrixClientCreds {
    homeserverUrl: string;
    identityServerUrl: string;
    userId: string;
    deviceId: string;
    accessToken: string;
    guest: boolean;
}

/**
 * Wrapper object for handling the js-sdk Matrix Client object in the react-sdk
 * Handles the creation/initialisation of client objects.
 * This module provides a singleton instance of this class so the 'current'
 * Matrix Client object is available easily.
 */
class MatrixClientPeg {
    constructor() {
        this.matrixClient = null;
        this._justRegisteredUserId = null;

        // These are the default options used when when the
        // client is started in 'start'. These can be altered
        // at any time up to after the 'will_start_client'
        // event is finished processing.
        this.opts = {
            initialSyncLimit: 20
        };
        // the credentials used to init the current client object.
        // used if we tear it down & recreate it with a different store
        this._currentClientCreds = null;
    }

    /**
     * Sets the script href passed to the IndexedDB web worker
     * If set, a separate web worker will be started to run the IndexedDB
     * queries on.
     *
     * @param {string} script href to the script to be passed to the web worker
     */
    setIndexedDbWorkerScript(script) {
        createMatrixClient.indexedDbWorkerScript = script;
    }

    get(): MatrixClient {
        return this.matrixClient;
    }

    unset() {
        this.matrixClient = null;

        MatrixActionCreators.stop();
    }

    /**
     * If we've registered a user ID we set this to the ID of the
     * user we've just registered. If they then go & log in, we
     * can send them to the welcome user (obviously this doesn't
     * guarentee they'll get a chat with the welcome user).
     *
     * @param {string} uid The user ID of the user we've just registered
     */
    setJustRegisteredUserId(uid) {
        this._justRegisteredUserId = uid;
    }

    /**
     * Returns true if the current user has just been registered by this
     * client as determined by setJustRegisteredUserId()
     *
     * @returns {bool} True if user has just been registered
     */
    currentUserIsJustRegistered() {
        return (
            this.matrixClient &&
            this.matrixClient.credentials.userId === this._justRegisteredUserId
        );
    }

    /*
     * Replace this MatrixClientPeg's client with a client instance that has
     * homeserver / identity server URLs and active credentials
     */
    replaceUsingCreds(creds: MatrixClientCreds) {
        this._currentClientCreds = creds;
        this._createClient(creds);
    }

    async assign() {
        for (const dbType of ["indexeddb", "memory"]) {
            try {
                const promise = this.matrixClient.store.startup();
                console.log(
                    "MatrixClientPeg: waiting for MatrixClient store to initialise"
                );
                await promise;
                break;
            } catch (err) {
                if (dbType === "indexeddb") {
                    console.error(
                        "Error starting matrixclient store - falling back to memory store",
                        err
                    );
                    this.matrixClient.store = new MemoryStore({
                        localStorage: global.localStorage
                    });
                } else {
                    console.error("Failed to start memory store!", err);
                    throw err;
                }
            }
        }

        StorageManager.trackStores(this.matrixClient);

        // try to initialise e2e on the new client
        try {
            // check that we have a version of the js-sdk which includes initCrypto
            if (
                !SettingsStore.getValue("lowBandwidth") &&
                this.matrixClient.initCrypto
            ) {
                await this.matrixClient.initCrypto();
                StorageManager.setCryptoInitialised(true);
            }
        } catch (e) {
            if (e && e.name === "InvalidCryptoStoreError") {
                // The js-sdk found a crypto DB too new for it to use
                const CryptoStoreTooNewDialog = sdk.getComponent(
                    "views.dialogs.CryptoStoreTooNewDialog"
                );
                Modal.createDialog(CryptoStoreTooNewDialog, {
                    host: window.location.host
                });
            }
            // this can happen for a number of reasons, the most likely being
            // that the olm library was missing. It's not fatal.
            console.warn("Unable to initialise e2e", e);
        }

        const opts = utils.deepCopy(this.opts);
        // the react sdk doesn't work without this, so don't allow
        opts.pendingEventOrdering = "detached";
        opts.lazyLoadMembers = true;

        // Connect the matrix client to the dispatcher and setting handlers
        MatrixActionCreators.start(this.matrixClient);
        MatrixClientBackedSettingsHandler.matrixClient = this.matrixClient;

        return opts;
    }

    async start() {
        const opts = await this.assign();

        console.log(`MatrixClientPeg: really starting MatrixClient`);
        await this.get().startClient(opts);
        console.log(`MatrixClientPeg: MatrixClient started`);
    }

    getCredentials(): MatrixClientCreds {
        return {
            homeserverUrl: this.matrixClient.baseUrl,
            identityServerUrl: this.matrixClient.idBaseUrl,
            userId: this.matrixClient.credentials.userId,
            deviceId: this.matrixClient.getDeviceId(),
            accessToken: this.matrixClient.getAccessToken(),
            guest: this.matrixClient.isGuest()
        };
    }

    /*
     * Return the server name of the user's homeserver
     * Throws an error if unable to deduce the homeserver name
     * (eg. if the user is not logged in)
     */
    getHomeserverName() {
        const matches = /^@.+:(.+)$/.exec(this.matrixClient.credentials.userId);
        if (matches === null || matches.length < 1) {
            throw new Error("Failed to derive homeserver name from user ID!");
        }
        return matches[1];
    }

    _createClient(creds: MatrixClientCreds) {
        const opts = {
            baseUrl: creds.homeserverUrl,
            idBaseUrl: creds.identityServerUrl,
            accessToken: creds.accessToken,
            userId: creds.userId,
            deviceId: creds.deviceId,
            timelineSupport: true,
            forceTURN: !SettingsStore.getValue("webRtcAllowPeerToPeer", false),
            fallbackICEServerAllowed: !!SettingsStore.getValue(
                "fallbackICEServerAllowed"
            ),
            verificationMethods: [verificationMethods.SAS],
            unstableClientRelationAggregation: true,
            identityServer: new IdentityAuthClient()
        };

        this.matrixClient = createMatrixClient(opts);

        // we're going to add eventlisteners for each matrix event tile, so the
        // potential number of event listeners is quite high.
        this.matrixClient.setMaxListeners(500);

        this.matrixClient.setGuest(Boolean(creds.guest));

        const notifTimelineSet = new EventTimelineSet(null, {
            timelineSupport: true
        });
        // XXX: what is our initial pagination token?! it somehow needs to be synchronised with /sync.
        notifTimelineSet
            .getLiveTimeline()
            .setPaginationToken("", EventTimeline.BACKWARDS);
        this.matrixClient.setNotifTimelineSet(notifTimelineSet);
        //console.log("OPTIONS TO GET CLIENT CREDENTIALS", opts);
    }
}

if (!global.mxMatrixClientPeg) {
    global.mxMatrixClientPeg = new MatrixClientPeg();
}
export default global.mxMatrixClientPeg;
