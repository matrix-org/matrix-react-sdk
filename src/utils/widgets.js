import MatrixClientPeg from '../MatrixClientPeg';

/**
 * Get all widgets (user and room) for the current user
 * @param  {object} room The room to get widgets for
 * @return {[object]} Array containing current / active room and user widget state events
 */
function getWidgets(room) {
    const widgets = getRoomWidgets(room);
    widgets.concat(getUserWidgetsArray());
    return widgets;
}

/**
 * Get room specific widgets
 * @param  {object} room The room to get widgets force
 * @return {[object]} Array containing current / active room widgets
 */
function getRoomWidgets(room) {
    const appsStateEvents = room.currentState.getStateEvents('im.vector.modular.widgets');
    if (!appsStateEvents) {
        return [];
    }

    return appsStateEvents.filter((ev) => {
        return ev.getContent().type && ev.getContent().url;
    });
}

/**
 * Get user specific widgets (not linked to a specific room)
 * @return {object} Event content object containing current / active user widgets
 */
function getUserWidgets() {
    const client = MatrixClientPeg.get();
    if (!client) {
        throw new Error('User not logged in');
    }
    const userWidgets = client.getAccountData('m.widgets');
    let userWidgetContent = {};
    if (userWidgets && userWidgets.getContent()) {
      userWidgetContent = userWidgets.getContent();
    }
    return userWidgetContent;
}

/**
 * Get user specific widgets (not linked to a specific room) as an array
 * @return {[object]} Array containing current / active user widgets
 */
function getUserWidgetsArray() {
  return Object.values(getUserWidgets());
}

/**
 * Get active stickerpicker widgets (stickerpickers are user widgets by nature)
 * @return {[object]} Array containing current / active stickerpicker widgets
 */
function getStickerpickerWidgets() {
    const widgets = getUserWidgetsArray();
    return widgets.filter((widget) => widget.content && widget.content.type === "m.stickerpicker");
}

/**
 * Remove all stickerpicker widgets (stickerpickers are user widgets by nature)
 * @return {Promise} Resolves on account data updated
 */
function removeStickerpickerWidgets() {
    const client = MatrixClientPeg.get();
    if (!client) {
        throw new Error('User not logged in');
    }
    const userWidgets = client.getAccountData('m.widgets').getContent() || {};
    Object.entries(userWidgets).forEach(([key, widget]) => {
        if (widget.content && widget.content.type === 'm.stickerpicker') {
            delete userWidgets[key];
        }
    });
    return client.setAccountData('m.widgets', userWidgets);
}


export default {
    getWidgets,
    getRoomWidgets,
    getUserWidgets,
    getUserWidgetsArray,
    getStickerpickerWidgets,
    removeStickerpickerWidgets,
};
