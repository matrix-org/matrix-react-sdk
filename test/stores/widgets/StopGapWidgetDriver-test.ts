/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { mocked, MockedObject } from "jest-mock";
import { Widget, WidgetKind, WidgetDriver, ITurnServer } from "matrix-widget-api";
import { MatrixClient, ClientEvent, ITurnServer as IClientTurnServer } from "matrix-js-sdk/src/client";
import { DeviceInfo } from "matrix-js-sdk/src/crypto/deviceinfo";

import { MatrixClientPeg } from "../../../src/MatrixClientPeg";
import { StopGapWidgetDriver } from "../../../src/stores/widgets/StopGapWidgetDriver";
import { stubClient } from "../../test-utils";

describe("StopGapWidgetDriver", () => {
    let client: MockedObject<MatrixClient>;
    let driver: WidgetDriver;

    beforeEach(() => {
        stubClient();
        client = mocked(MatrixClientPeg.get());

        driver = new StopGapWidgetDriver(
            [],
            new Widget({
                id: "test",
                creatorUserId: "@alice:example.org",
                type: "example",
                url: "https://example.org",
            }),
            WidgetKind.Room,
        );
    });

    describe("sendToDevice", () => {
        const contentMap = {
            "@alice:example.org": {
                "*": {
                    hello: "alice",
                },
            },
            "@bob:example.org": {
                "bobDesktop": {
                    hello: "bob",
                },
            },
        };

        it("sends unencrypted messages", async () => {
            await driver.sendToDevice("org.example.foo", false, contentMap);
            expect(client.queueToDevice.mock.calls).toMatchSnapshot();
        });

        it("sends encrypted messages", async () => {
            const aliceWeb = new DeviceInfo("aliceWeb");
            const aliceMobile = new DeviceInfo("aliceMobile");
            const bobDesktop = new DeviceInfo("bobDesktop");

            mocked(client.crypto.deviceList).downloadKeys.mockResolvedValue({
                "@alice:example.org": { aliceWeb, aliceMobile },
                "@bob:example.org": { bobDesktop },
            });

            await driver.sendToDevice("org.example.foo", true, contentMap);
            expect(client.encryptAndSendToDevices.mock.calls).toMatchSnapshot();
        });
    });

    describe("getTurnServers", () => {
        it("stops if VoIP isn't supported", async () => {
            jest.spyOn(client, "pollingTurnServers", "get").mockReturnValue(false);
            const servers = driver.getTurnServers();
            expect(await servers.next()).toEqual({ value: undefined, done: true });
        });

        it("stops if the homeserver provides no TURN servers", async () => {
            const servers = driver.getTurnServers();
            expect(await servers.next()).toEqual({ value: undefined, done: true });
        });

        it("gets TURN servers", async () => {
            const server1: ITurnServer = {
                uris: [
                    "turn:turn.example.com:3478?transport=udp",
                    "turn:10.20.30.40:3478?transport=tcp",
                    "turns:10.20.30.40:443?transport=tcp",
                ],
                username: "1443779631:@user:example.com",
                password: "JlKfBy1QwLrO20385QyAtEyIv0=",
            };
            const server2: ITurnServer = {
                uris: [
                    "turn:turn.example.com:3478?transport=udp",
                    "turn:10.20.30.40:3478?transport=tcp",
                    "turns:10.20.30.40:443?transport=tcp",
                ],
                username: "1448999322:@user:example.com",
                password: "hunter2",
            };
            const clientServer1: IClientTurnServer = {
                urls: server1.uris,
                username: server1.username,
                credential: server1.password,
            };
            const clientServer2: IClientTurnServer = {
                urls: server2.uris,
                username: server2.username,
                credential: server2.password,
            };

            client.getTurnServers.mockReturnValue([clientServer1]);
            const servers = driver.getTurnServers();
            expect(await servers.next()).toEqual({ value: server1, done: false });

            const nextServer = servers.next();
            client.getTurnServers.mockReturnValue([clientServer2]);
            client.emit(ClientEvent.TurnServers, [clientServer2]);
            expect(await nextServer).toEqual({ value: server2, done: false });

            await servers.return(undefined);
        });
    });
});
