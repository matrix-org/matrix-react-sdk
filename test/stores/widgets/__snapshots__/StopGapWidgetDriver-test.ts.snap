// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StopGapWidgetDriver sendToDevice sends encrypted messages 1`] = `
[
  [
    [
      {
        "deviceInfo": DeviceInfo {
          "algorithms": [],
          "deviceId": "aliceWeb",
          "keys": {},
          "known": false,
          "signatures": {},
          "unsigned": {},
          "verified": 0,
        },
        "userId": "@alice:example.org",
      },
      {
        "deviceInfo": DeviceInfo {
          "algorithms": [],
          "deviceId": "aliceMobile",
          "keys": {},
          "known": false,
          "signatures": {},
          "unsigned": {},
          "verified": 0,
        },
        "userId": "@alice:example.org",
      },
    ],
    {
      "hello": "alice",
    },
  ],
  [
    [
      {
        "deviceInfo": DeviceInfo {
          "algorithms": [],
          "deviceId": "bobDesktop",
          "keys": {},
          "known": false,
          "signatures": {},
          "unsigned": {},
          "verified": 0,
        },
        "userId": "@bob:example.org",
      },
    ],
    {
      "hello": "bob",
    },
  ],
]
`;

exports[`StopGapWidgetDriver sendToDevice sends unencrypted messages 1`] = `
[
  [
    {
      "batch": [
        {
          "deviceId": "*",
          "payload": {
            "hello": "alice",
          },
          "userId": "@alice:example.org",
        },
        {
          "deviceId": "bobDesktop",
          "payload": {
            "hello": "bob",
          },
          "userId": "@bob:example.org",
        },
      ],
      "eventType": "org.example.foo",
    },
  ],
]
`;
